runtime/src/bank.rs

```diff
@@ -4414,29 +4435,33 @@ impl Bank {
 
         // parallelize?
         let rent_for_sysvars = self.rent_for_sysvars();
-        let mut total_rent = 0;
         let mut rent_debits = RentDebits::default();
+        let mut total_collected = CollectedInfo::default();
         for (pubkey, mut account) in accounts {
-            let rent = self.rent_collector.collect_from_existing_account(
+            let collected = self.rent_collector.collect_from_existing_account(
                 &pubkey,
                 &mut account,
                 rent_for_sysvars,
                 self.rc.accounts.accounts_db.filler_account_suffix.as_ref(),
             );
-            total_rent += rent;
+            total_collected += collected;
             // Store all of them unconditionally to purge old AppendVec,
             // even if collected rent is 0 (= not updated).
             // Also, there's another subtle side-effect from this: this
             // ensures we verify the whole on-chain state (= all accounts)
             // via the account delta hash slowly once per an epoch.
             self.store_account(&pubkey, &account);
-            rent_debits.insert(&pubkey, rent, account.lamports());
+            rent_debits.insert(&pubkey, collected.rent_amount, account.lamports());
         }
-        self.collected_rent.fetch_add(total_rent, Relaxed);
+        self.collected_rent
+            .fetch_add(total_collected.rent_amount, Relaxed);
         self.rewards
             .write()
             .unwrap()
             .extend(rent_debits.into_unordered_rewards_iter());
+        if total_collected.account_data_len > 0 {
+            self.update_accounts_data_len(-(total_collected.account_data_len as i64));
```

Comment:
Originally I wanted to use atomic `fetch_sub()` here, but that doesn't have a 'saturating' version (more details below). I need the `update_accounts_data_len()` version for a future PR to handle where the accounts data len can either shrink or grow, so pulled that in to use here.

So `fetch_sub()` _should_ work in all real/production cases†, but causes quite a few tests to fail. This is because the Bank that's created for tests has an `accounts_data_len` of 0. The test adds accounts, but not in a way that updates accounts data len. Thus, the tests that check for rent cleaning up accounts cause the `accounts_data_len` to wrap back to a very large positive number.

† The bank's accounts_data_len is not 100% the true size, and I'm not sure it will be. So I'm also concerned of a production scenario that causes the wrap around to happen and takes down the cluster.

To solve, I could've:
1. Update all the test by manually setting a non zero `accounts_data_len`
2. Have a test and not-test (`#[cfg(test)]` and `#[cfg(not(test))]` block right here at this line
3. My current impl

I went with (3) because (1) seems fragile, and (2) seems a bit dangerous to not have the same logic for tests and not tests on such a core component.

Let me know if you'd like a different solution!